Implement a monotonic cubic interpolator with at least C1 continuity

lindsayad opened the pull request:
Algorithm is based on Fritsch's and Carlson's 1980 paper "Monotone Piecewise Cubic Interpolation." In brief, the algorithm is:

1. From input data set, initialize derivatives using second order accurate differencing
2.  Check derivatives against a region of monotonicity. 
3. If outside monotonicity region, modify derivatives accordingly.

References issue #7158

Commit:
Add MonotoneCubicInterpolation class (refs #7158).

Commit:
Add sampleDerivative method. (#7158)

Commit:
Debug and add test. (#7158)

lindsayad commented:
Still to do: 
 - Add sample2ndDerivative
 - squash small commits

permcody reviewed (commented):
This looks pretty good. I'll make a general comment about these utilities. I'd like to see them separated a little more from MOOSE. Some of them are, some of them aren't. LinearInterpolation is one of the better ones. It throws errors instead of calling MOOSE errors and is a stand alone utility. It even has tests in the unit suite. Some of these other utilities are linked into MOOSE (call mooseError() for instance) and only have full integration tests.

jwpeterson commented on framework/include/utils/MonotoneCubicInterpolation.h:
```diff
@@ -0,0 +1,82 @@
+
+  virtual ~MonotoneCubicInterpolation() = default;
+
+  virtual void setData(const std::vector<Real> x, const std::vector<Real> y);
```
`x` and `y` should be passed by const reference.

jwpeterson commented on framework/include/utils/MonotoneCubicInterpolation.h:
```diff
@@ -0,0 +1,82 @@
+  virtual void setData(const std::vector<Real> x, const std::vector<Real> y);
+  virtual Real sample(Real x) const;
+  virtual Real sampleDerivative(Real x) const;
+  virtual void dumpCSV(std::string filename, const std::vector<Real> xnew);
```
`xnew` should be passed by const reference.

jwpeterson commented on framework/include/utils/MonotoneCubicInterpolation.h:
```diff
@@ -0,0 +1,82 @@
+  virtual Real p(const Real & xhi, const Real & xlo, const Real & fhi, const Real & flo,
+                 const Real & dhi, const Real & dlo, const Real & x) const;
+  virtual Real pPrime(const Real & xhi, const Real & xlo, const Real & fhi, const Real & flo,
+                 const Real & dhi, const Real & dlo, const Real & x) const;
```
Indentation does not line up with line above.

jwpeterson commented on framework/include/utils/MonotoneCubicInterpolation.h:
```diff
@@ -0,0 +1,82 @@
+{
+public:
+  MonotoneCubicInterpolation();
+  MonotoneCubicInterpolation(const std::vector<Real> x, const std::vector<Real> y);
```
`x` and `y` should be passed by const reference.

jwpeterson commented on framework/src/utils/MonotoneCubicInterpolation.C:
```diff
@@ -0,0 +1,299 @@
+Real
+MonotoneCubicInterpolation::sign(Real x) const
+{
+  return x < 0 ? -1. : (x > 0 ? 1. : 0.);
```
![xz](https://cloud.githubusercontent.com/assets/1775907/21018595/a5cbae1a-bd2a-11e6-8a83-e208b9f49abc.jpg)

jwpeterson commented on framework/src/utils/MonotoneCubicInterpolation.C:
```diff
@@ -0,0 +1,299 @@
+      return;
+    }
+  }
+  s > 0 ? _monotonic_e = "monotonic_increase" :
```
Please just use normal if-statements, it's much easier to read.

jwpeterson commented on framework/src/utils/MonotoneCubicInterpolation.C:
```diff
@@ -0,0 +1,299 @@
+Real
+MonotoneCubicInterpolation::phi(const Real & t) const
+{
+  return 3. * std::pow(t, 2) - 2. * std::pow(t, 3);
```
`t*t` and `t*t*t` are shorter and just as readable as `std::pow` calls IMO.

Commit:
Incorporate John suggestions. Add sample2ndDerivative. More efficient interval find (#7158).

Commit:
Fix end point conditions (#7158).

Commit:
Remove MOOSE from MonotoneCubicInterpolation. (#7158)

Commit:
Create unit tests. (#7158)

Commit:
Remove integration test. (#7158)

lindsayad commented:
Hopefully I addressed your review @jwpeterson. @permcody I migrated my tests over to unit and removed all dependence of the interpolation utility on MOOSE.

permcody commented:
Thanks @lindsayad, this looks much better!

permcody commented:
@brianmoose - Thanks for fixing up the "developer activated logic". Please add @lindsayad to the approved group (Moose Collaborators?).

dschwen commented:
Why does this use `double` all over the place?

lindsayad commented:
Instead of `Real`? Based off @permcody's recommendation, I divorced this class completely from libmesh/MOOSE such that it should be able to be a stand-alone utility.

dschwen commented:
I don't think that is a great idea. It is in the MOOSE framework directory. There is a (small) risk that we run into cases where Real != double. And in my opinion (and that might sound a bit harsh ;-) ) this _pollutes_ the MOOSE code in which we very consistently use Real (unless we are coupling to external libraries or fortran codes).

lindsayad commented:
I don't take any offense. This is a difference of opinion that I don't care so much about. Whether the team wants it to be a completely stand-alone utility or whether it should maintain consistency with Real, it's only a `sed` command away. I'll abide by your opinion.

lindsayad commented:
I will say that the `SplineInterpolation` utility used `double` until I came and re-wrote it in order to add the `BicubicSpline` and switched all occurrences with `Real` ;-)

permcody commented:
I'll chime in. I'd actually prefer using `Real` too. I'm very happy that
you pulled this out of MOOSE as requested but it's possible to still
include the header with the typedef to use `Real`. It can be added back in
if there are people that would like to see it that way.

On Wed, Dec 14, 2016 at 12:28 PM Alex Lindsay <notifications@github.com>
wrote:

> I will say that the SplineInterpolation utility used double until I came
> and re-wrote it in order to add the BicubicSpline and switched all
> occurrences with Real ;-)
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/8193#issuecomment-267131242>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AC5XIJE3N_MzPqA0l8wUdBioOHrSQl6Yks5rIENbgaJpZM4LHZAk>
> .
>

Commit:
Swap double back to Real. (#7158)

lindsayad commented:
Alrighty, double swapped back to Real

lindsayad commented:
Thanks @jwpeterson!

jwpeterson commented:
No problem.  After merging, I realized there's not really any documentation for the new class.  Would you mind adding a Doxygen-style comment block to the top of the class, i.e.
```
/**
 * Blah blah blah
 */
```
which gives the reference to Fritsch and Carlson's paper, and document any parts of the public interface which users are supposed to call.

lindsayad commented:
Yea, I'll get on it.

On 12/19/2016 09:31 AM, John W. Peterson wrote:
>
> No problem. After merging, I realized there's not really any 
> documentation for the new class. Would you mind adding a Doxygen-style 
> comment block to the top of the class, i.e.
>
> |/** * Blah blah blah */ |
>
> which gives the reference to Fritsch and Carlson's paper, and document 
> any parts of the public interface which users are supposed to call.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://github.com/idaholab/moose/pull/8193#issuecomment-267993586>, 
> or mute the thread 
> <https://github.com/notifications/unsubscribe-auth/AJxgcDvu5O0RkgymPh6hylKz6cj-mY0hks5rJqNSgaJpZM4LHZAk>.
>
